bundle/src/main/java/com/adobe/acs/commons/reports/models/QueryReportExecutor.java

```diff
@@ -45,165 +46,242 @@
+            if (isQueryBuilder()) {
+                results = getResultsFromQueryBuilder(limit, offset);
+            } else {
+                results = getResultsFromQuery(limit, offset);
+            }
+            return new ResultsPage(results.getLeft(), config.getPageSize(), page, results.getRight());
+        } catch (RepositoryException re) {
+            throw new ReportException("Exception executing search results", re);
+        }
+    }
 
-  public QueryReportExecutor(SlingHttpServletRequest request) {
-    this.request = request;
-  }
+    private boolean isQueryBuilder() {
+        return QueryHelperImpl.QUERY_BUILDER.equals(config.getQueryLanguage());
+    }
+
+    private Session getSession() throws ReportException {
+        return Optional.ofNullable(request.getResourceResolver().adaptTo(Session.class))
+                .orElseThrow(() -> new ReportException("Failed to get JCR Session"));
+    }
 
-  private ResultsPage fetchResults(int limit, int offset) throws ReportException {
-    prepareStatement();
-    try {
-      ResourceResolver resolver = request.getResourceResolver();
-      QueryManager queryMgr = Optional.ofNullable(resolver.adaptTo(Session.class))
-          .orElseThrow(() -> new ReportException("Failed to get JCR Session")).getWorkspace().getQueryManager();
+    private com.day.cq.search.Query prepareQueryBuilderQuery() throws ReportException {
 
-      Query query = queryMgr.createQuery(statement, config.getQueryLanguage());
+        Session session = getSession();
+        final Map<String, String> params = ParameterUtil.toMap(prepareStatement().split("\n\n?"), "=", false, null,
+                true);
 
-      if (page != -1) {
-        log.debug("Fetching results with limit {} and offset {}", limit, offset);
-        query.setLimit(limit);
-        query.setOffset(offset);
-      } else {
-        log.debug("Fetching all results");
-      }
-      QueryResult result = query.execute();
+        return queryBuilder.createQuery(PredicateGroup.create(params),
+                session);
+    }
 
-      NodeIterator nodes = result.getNodes();
+    private Pair<Stream<Resource>, Long> getResultsFromQueryBuilder(
+            int limit, int offset) throws ReportException {
 
-      Spliterator<Node> spliterator = Spliterators.spliteratorUnknownSize(nodes,
-          Spliterator.ORDERED | Spliterator.NONNULL);
+        com.day.cq.search.Query query = prepareQueryBuilderQuery();
+        if (page != -1) {
+            log.debug("Fetching results with limit {} and offset {}", limit, offset);
+            query.setHitsPerPage(limit);
+            query.setStart(offset);
+        } else {
+            log.debug("Fetching all results");
+        }
 
-      Stream<Resource> results = StreamSupport.stream(spliterator, false).map(n -> getResource(n, resolver));
+        SearchResult result = query.getResult();
+        Spliterator<Resource> spliterator = Spliterators.spliteratorUnknownSize(result.getResources(),
```

Comment:
Do we have to worry about the leaking ResourceResolver in QB?

https://github.com/Adobe-Consulting-Services/acs-aem-samples/blob/master/core/src/main/java/com/adobe/acs/samples/search/querybuilder/impl/SampleQueryBuilder.java#L159-L198

Reply:
@davidjgonzalez @klcodanr we've got this utility method that prevents the Query from creating its own ResourceResolver: https://github.com/Adobe-Consulting-Services/acs-aem-commons/blob/06d2c1c5e82545feabba8251e6f3a0ee503964e4/bundle/src/main/java/com/adobe/acs/commons/cqsearch/QueryUtil.java#L57

Reply:
I took a different approach to map the hits into resources rather than calling the getResource method. As far as I can tell this avoid creating the resource resolver since it's lazily created.

Reply:
@klcodanr Ah, I see what you did there 👍